README.md
# RepeatExplorer2 with TAREAN (Tandem Repeat Analyzer) #
-------------------------------------------------------------------------------
New version of RepeatExplorer with TAndem REpeat ANalyzer 

## Authors
Petr Novak, Jiri Macas, Pavel Neumann
Biology Centre CAS, Czech Republic


## How to use RepeatExplorer2

We recommend to use RepeatExplorer2 through Galaxy based web interface on our
public server at address
[https://galaxy-elixir.cerit-sc.cz](https://galaxy-elixir.cerit-sc.cz). This
server is provided as part of the [Elixir CZ project](https://www.elixir-czech.cz) and is maintained by CESNET
and CERIT-SC that are participants of this project. Server provide sufficient
computational resources for full scale RepeatExplorer2 or TAREAN analysis. This
Galaxy server also include all necessary tools for data preprocessing and QC and
some additional tools for repeat annotation.

Step by step protocols how to use RepeatExplorer on Galaxy server were published
in [Nature Protocols](http://dx.doi.org/10.1038/s41596-020-0400-y)

Free link to the full-text: [https://rdcu.be/b80Gr](https://rdcu.be/b80Gr)

Four protocols are included:

1. *De novo* repeat identification in a single species
2. Comparative repeat analysis in a set of species
3. Development of satellite DNA probes for cytogenetic experiments
4. Identification of centromeric repeats based on ChIP-seq data


If you prefer to use command line version of RepeatExplorer2 and TAREAN, read the instruction below. 

## Change log
this repository was trasfered from bitbucket (https://bitbucket.org/petrnovak/repex_tarean/) on 2023-11-30. 
[link](CHANGELOG.md)


  
## Installation ##
To use RepeatExplorer without installation, we recommend our freely
available Galaxy server at
[https://repeatexplorer-elixir.cerit-sc.cz](https://repeatexplorer-elixir.cerit-sc.cz).
This server is provided in frame of the ELIXIR-CZ project. The Galaxy
server includes also additional tools useful for data preprocessing, quality control
and genome annotation.

For installing command-line version, follow the instruction below:


Download source code using git command:

	git clone https://bitbucket.org/petrnovak/repex_tarean.git
	cd repex_tarean
    
We recommend to install dependencies using conda (conda can be installed using [miniconda](https://docs.conda.io/en/latest/miniconda.html)). The required environment can be prepared using command:

    conda env create -f environment.yml

Then activate the prepared environment using:
   
    conda activate repeatexplorer

In the `repex_tarean` direcory compile the source and prepare databases using:

	make

Support for 32-bit executables is required. If you are using Ubuntu distribution you can add 32-bit support by running:

    sudo dpkg --add-architecture i386
    sudo apt-get update
    sudo apt-get install libc6:i386 libncurses5:i386 libstdc++6:i386
   

To verify the installation, run RepeatExplorer clustering on the provided example data:

    ./seqclust -p -v tmp/clustering_output test_data/LAS_paired_10k.fas
    

### Running RepeatExplorer using Singularity:
Singularity container is available from singularity hub. Multiple versions of
repeatexplorer are available. See [singularity](https://github.com/repeatexplorer/repex_tarean) conainted definiton for
available versions. To run latest version use:
`shub://repeatexplorer/repex_tarean` or
`shub://repeatexplorer/repex_tarean:latest`. To use specific version use e.g.
shub://repeatexplorer/repex_tarean:0.3.8.dbaa07f

Example of running RepeatExplorer2 using singularity:

get help:

    singularity exec shub://repeatexplorer/repex_tarean seqclust --help


run clustering in working directory:


    mkdir working_dir && cd working_dir
    # get test data - fasta file with paired reads in interlaced format
    wget https://bitbucket.org/petrnovak/repex_tarean/raw/devel/test_data/LAS_paired_10k.fas
    singularity exec --bind ${PWD}:/data/ shub://repeatexplorer/repex_tarean seqclust -p -v /data/re_output /data/LAS_paired_10k.fas
    
    
output will be locate in `working_dir/re_output`



### Running RepeatExplorer using Docker container
Singularity is prefered option but RepeatExplorer2 can be also run using Docker container((https://hub.docker.com/repository/docker/kavonrtep/repeatexplorer)) using following commands:


    # see help:
    docker run kavonrtep/repeatexplorer:2.3.8 /repex_tarean/seqclust --help
    # create working directory
    mkdir working_dir && cd working_dir
    # get test data - fasta file withj paired reads in interlaced format
    wget https://bitbucket.org/petrnovak/repex_tarean/raw/devel/test_data/LAS_paired_10k.fas
    # run clustering on test data in working directory
    docker run -v  ${PWD}:/data kavonrtep/repeatexplorer:2.3.8 /repex_tarean/seqclust -p -v /data/re_output /data/LAS_paired_10k.fas
    # output will be locate in working_dir/re_output 




## Protein databases

Repeatexplorer2 utilizes REXdb database of protein domains for repeat annotation and classification. Structure of the database is described at [http://repeatexplorer.org/](http://repeatexplorer.org/). REXdb is distributed together with RepeatExplorer source code (`database` directory)


## RepeatExplorer command line options

    usage: seqclust [-h] [-p] [-A] [-t] [-l LOGFILE] [-m {float range 0.0..100.0}]
                    [-M {0,float range 0.1..1}] [-o {float range 30.0..80.0}]
                    [-c CPU] [-s SAMPLE] [-P PREFIX_LENGTH] [-v OUTPUT_DIR]
                    [-r MAX_MEMORY] [-d DATABASE DATABASE] [-C] [-k]
                    [-a {2,3,4,5}]
                    [-tax {VIRIDIPLANTAE3.0,VIRIDIPLANTAE2.2,METAZOA2.0,METAZOA3.0}]
                    [-opt {ILLUMINA,ILLUMINA_DUST_OFF,ILLUMINA_SHORT,OXFORD_NANOPORE}]
                    [-D {BLASTX_W2,BLASTX_W3,DIAMOND}]
                    sequences
    
    RepeatExplorer:
        Repetitive sequence discovery and clasification from NGS data
    
        
    
    positional arguments:
      sequences
    
    optional arguments:
      -h, --help            show this help message and exit
      -p, --paired
      -A, --automatic_filtering
      -t, --tarean_mode     analyze only tandem reapeats without additional classification
      -l LOGFILE, --logfile LOGFILE
                            log file, logging goes to stdout if not defines
      -m {float range 0.0..100.0}, --mincl {float range 0.0..100.0}
      -M {0,float range 0.1..1}, --merge_threshold {0,float range 0.1..1}
                            threshold for mate-pair based cluster merging, default 0 - no merging
      -o {float range 30.0..80.0}, --min_lcov {float range 30.0..80.0}
                            minimal overlap coverage - relative to longer sequence length, default 55
      -c CPU, --cpu CPU     number of cpu to use, if 0 use max available
      -s SAMPLE, --sample SAMPLE
                            use only sample of input data[by default max reads is used
      -P PREFIX_LENGTH, --prefix_length PREFIX_LENGTH
                            If you wish to keep part of the sequences name,
                             enter the number of characters which should be 
                            kept (1-10) instead of zero. Use this setting if
                             you are doing comparative analysis
      -v OUTPUT_DIR, --output_dir OUTPUT_DIR
      -r MAX_MEMORY, --max_memory MAX_MEMORY
                            Maximal amount of available RAM in kB if not set
                            clustering tries to use whole available RAM
      -d DATABASE DATABASE, --database DATABASE DATABASE
                            fasta file with database for annotation and name of database
      -C, --cleanup         remove unncessary large files from working directory
      -k, --keep_names      keep sequence names, by default sequences are renamed
      -a {2,3,4,5}, --assembly_min {2,3,4,5}
                            Assembly is performed on individual clusters, by default 
                            clusters with size less then 5 are not assembled. If you 
                            want need assembly of smaller cluster set *assmbly_min* 
                            accordingly
      -tax {VIRIDIPLANTAE3.0,VIRIDIPLANTAE2.2,METAZOA2.0,METAZOA3.0}, --taxon {VIRIDIPLANTAE3.0,VIRIDIPLANTAE2.2,METAZOA2.0,METAZOA3.0}
                            Select taxon and protein database version
      -opt {ILLUMINA,ILLUMINA_DUST_OFF,ILLUMINA_SHORT,OXFORD_NANOPORE}, --options {ILLUMINA,ILLUMINA_DUST_OFF,ILLUMINA_SHORT,OXFORD_NANOPORE}
      -D {BLASTX_W2,BLASTX_W3,DIAMOND}, --domain_search {BLASTX_W2,BLASTX_W3,DIAMOND}
                            Detection of protein domains can be performed by either blastx or
                             diamond" program. options are:
                              BLASTX_W2 - blastx with word size 2 (slowest, the most sesitive)
                              BLASTX_W3 - blastx with word size 3 (default)
                              DIAMOND   - diamond program (significantly faster, less sensitive)
                            To use this option diamond program must be installed in your PATH
    


## Galaxy toolshed

RepeatExplorer can be installed on Galaxy server from toolshed repository [ https://toolshed.g2.bx.psu.edu/view/petr-novak/repeatexplorer2]( https://toolshed.g2.bx.psu.edu/view/petr-novak/repeatexplorer2).

## Reproducibility
To make clustering reproducible between runs with the 
same data, environment variable PYTHONHASHSEED must be set:

	export PYTHONHASHSEED=0
    
## Disk space requirements
Large sqlite database for temporal data is created in OS-specific temp directory- usually /tmp/ 
To use alternative location, it is necessary to specify `TEMP` environment variable.

## CPU and RAM requirements

Resources requirements can be set either from command line arguments `--max-memory` and `--cpu` or
using environment variables `TAREAN_MAX_MEM` and `TAREAN_CPU`. If not set, the pipeline will use all
available resources

## How to cite

If you use RepeatExplorer for general repeat characterization in your work please cite:

 - [Novak, P., Neumann, P., Pech, J., Steinhaisl, J., Macas, J. (2013) - RepeatExplorer: a Galaxy-based web server for genome-wide characterization of eukaryotic repetitive elements from next generation sequence read. Bioinformatics 29:792-793](http://bioinformatics.oxfordjournals.org/content/29/6/792)

or

 - [Novak, P., Neumann, P., Macas, J. (2010) - Graph-based clustering and characterization of repetitive sequences in next-generation sequencing data. BMC Bioinformatics 11 :37](http://www.biomedcentral.com/1471-2105/11/378)

If you use TAREAN for satellite detection and characterization please cite:

 - [Novak, P., Robledillo, L.A.,Koblizkova, A., Vrbova, I., Neumann, P., Macas, J. (2017) - TAREAN: a computational tool for identification and characterization of satellite DNA from unassembled short reads. Nucleic Acid Research](https://doi.org/10.1093/nar/gkx257)
 


CHANGELOG.md
# RepeatExplorer2 Changelog

## 0.3.8.1 (2.3.8.1 version on Galaxy)
May 27, 2022

   - additional output in comparative analysis - observer/expected edges ratio is now available as CSV file

## 0.3.8. (2.3.8. version on Galaxy)
Apr 9 2020

   - Some additional output added  - links to fasta and csv files in HTML report
   - Sensistive all-to-all option added

## 0.3.7 (2.3.7 version on Galaxy)
Jan 9 2020

   - script for preparation of toolshed package added
   - RepeatExplorer2 and TAREAN are available also as Galaxy toolshed package: https://toolshed.g2.bx.psu.edu/view/petr-novak/repeatexplorer2
   - better version reporting 

##  0.3.6
Dec 20 2019
  
   - some dependencies removed 
   - all dependencies can be installed using miniconda

## v0.3.5
Nov 22 2019

   - better reporting when filtering of abundant satellite is used
   - updated format of HTML report
   - number of changes in names of output files - file names are now more informative. Changes include:
   
| old file name                                                            | new file name                                                        |
|--------------------------------------------------------------------------|----------------------------------------------------------------------|
| TR_consensus_rank_1_.fasta                                               | TAREAN_consensus_rank_1.fasta                                        |
| TR_consensus_rank_2_.fasta                                               | TAREAN_consensus_rank_2.fasta                                        |
| TR_consensus_rank_3_.fasta                                               | TAREAN_consensus_rank_3.fasta                                        |
| TR_consensus_rank_4_.fasta                                               | TAREAN_consensus_rank_4.fasta                                        |
| sequences/sequences.fasta                                                | reads/reads.fasta                                                    |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln.info.minRD5_sort-GR     | clustering/clusters/dir_CL0001/contigs.info.minRD5_sort-GR.fasta     |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln.info.minRD5_sort-length | clustering/clusters/dir_CL0001/contigs.info.minRD5_sort-length.fasta |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln.info.minRD5             | clustering/clusters/dir_CL0001/contigs.info.minRD5_sort-RD.fasta     |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln.profile                 | clustering/clusters/dir_CL0001/contigs.profile                       |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln.info                    | clustering/clusters/dir_CL0001/contigs.info.fasta                    |
| clustering/clusters/dir_CL0001/reads.fas.CL1.ace                         | clustering/clusters/dir_CL0001/contigs.ace                           |
| clustering/clusters/dir_CL0001/reads.fas.CL1.contigs.qual                | clustering/clusters/dir_CL0001/contigs.qual                          |
| clustering/clusters/dir_CL0001/reads.fas.CL1.aln                         | clustering/clusters/dir_CL0001/contigs.aln                           |
| clustering/clusters/dir_CL0001/reads.fas.CL1.contigs                     | clustering/clusters/dir_CL0001/contigs.fasta                         |
| clustering/clusters/dir_CL0001/reads.fas.CL1.info                        | clustering/clusters/dir_CL0001/assembly.info                         |
| clustering/clusters/dir_CL0001/reads.fas.CL1.singlets                    | clustering/clusters/dir_CL0001/singlets.fasta                        |
| clustering/clusters/dir_CL0001/reads.fas.CL1.contigs.links               | clustering/clusters/dir_CL0001/contigs.links                         |
| clustering/clusters/dir_CL0001/reads_oriented.fas                        | clustering/clusters/dir_CL0001/reads_selection_oriented.fasta        |
| clustering/clusters/dir_CL0001/CL1_directed_graph.RData.                 | clustering/clusters/dir_CL0001/graph_layout_directed.RData           |
| clustering/clusters/dir_CL0001/CL1_tmb.png                               | clustering/clusters/dir_CL0001/graph_layout_tmb.png                  |
| clustering/clusters/dir_CL0001/CL1.png                                   | clustering/clusters/dir_CL0001/graph_layout.png                      |
| clustering/clusters/dir_CL0001/CL1.GL                                    | clustering/clusters/dir_CL0001/graph_layout.GL                       |
| clustering/clusters/dir_CL0001/blast.csv                                 | clustering/clusters/dir_CL0001/hitsort_part.csv                      |

## v0.3.4 (1.0.0 version on Galaxy)
Oct 31 2019

  - Classification of superclusters improved, classification now uses information about LTR/PBS for classification of Class_I elements
  

## v0.3.2
Oct 9 2019

  - Graphical reporting of comparative analysis added 

## v0.3.1
Jan 9 2019

  - Improved detection of low complexity repeats and satellites with shorter monomer

## v0.3.0
Oct 25 2018

 - For back-compatibility, it is possible to select protein database version
 - Databases of protein domains went public 

## v0.2.10
Oct 24 2018

 - Protein database for Viridiplantae updated

## v0.2.9
Jan 3 2018

 - read depth for contigs is calculated
 - contigs are also sorted based on genome representation
 
## v0.2.8
Dec 18 2017

 - by default assembly is done on clusters with size at least 5 reads
 - preset option of Illumina, Illumina short and Oxford Nanopore
 - option for analyzing Metazoa/ Viridiplantae added
 - protein databases can be obtained **using fetch_databases.sh** script (password required)

## v0.2.7
Dec 05 2017

 - improved DNA database - protein domain were masked not to interfere with protein domain database
 - another bug fix in parallelization and assembly 
 - alternative to blastx added - DIAMOND program can be used instead
 
## v0.2.6
Nov 14 2017


config.py
'''
All configuration for clustering
'''
import os
import tempfile
from math import exp
from collections import namedtuple
MAIN_DIR = os.path.dirname(os.path.realpath(__file__))
def add_base_path(base):
    '''automates generating absolute path in config'''
    def joined_path(p):
        '''create absolute path function '''
        return os.path.join(base, p)
    return joined_path

PATH = add_base_path(MAIN_DIR)

# clustering general settings
DIRECTORY_TREE = {'libdir': 'libdir',
                  'seqclust': 'seqclust',
                  'assembly': 'seqclust/small_clusters_assembly',
                  'blastx': 'seqclust/blastx',
                  'clustering': 'seqclust/clustering',
                  'clusters': 'seqclust/clustering/clusters',
                  'superclusters': 'seqclust/clustering/superclusters',
                  'mgblast': 'seqclust/mgblast',
                  'blastn': 'seqclust/blastn',
                  'prerun': 'seqclust/prerun',
                  'prerun_clusters': 'seqclust/prerun/clusters',
                  'sequences': 'seqclust/reads',
                  'custom_databases': 'seqclust/custom_databases'}

if "TEMP" in os.environ:
    DIRECTORY_TREE['TEMP'] = os.environ["TEMP"]
else:
    DIRECTORY_TREE['TEMP'] = tempfile.TemporaryDirectory().name

FILES = {'sample_db': DIRECTORY_TREE['TEMP'] + "/sample.db",
         'sample_fasta': DIRECTORY_TREE['prerun'] + "/sample.fasta",
         'prerun_cls_file' : DIRECTORY_TREE['prerun'] + "/sample_hitsort.cls",
         'filter_sequences_file' : DIRECTORY_TREE['prerun'] + "/filter_sequences.fasta",
         'sequences_db': DIRECTORY_TREE['TEMP'] + "/sequences.db",
         'sequences_fasta': DIRECTORY_TREE['sequences'] + "/reads.fasta",
         'hitsort': DIRECTORY_TREE['clustering'] + "/hitsort",
         'hitsort_db': DIRECTORY_TREE['TEMP'] + "/hitsort.db",
         'cls_file': DIRECTORY_TREE['clustering'] + "/hitsort.cls",
         'clusters_summary_csv': "CLUSTER_TABLE.csv",
         'profrep_classification_csv': "PROFREP_CLASSIFICATION_TEMPLATE.csv",
         'superclusters_csv_summary': "SUPERCLUSTER_TABLE.csv",
         'comparative_analysis_counts_csv': "COMPARATIVE_ANALYSIS_COUNTS.csv",
         'clusters_info': ".clusters_info.csv",
         'tarean_report_html': "tarean_report.html",
         'cluster_report_html' : "cluster_report.html",
         'supercluster_report_html' : 'supercluster_report.html',
         'repeat_annotation_summary_rds' : 'repeat_annotation_summary.rds',
         'summarized_annotation_html' :'summarized_annotation.html',
         'main_report_html' : 'index.html',
         'TR_consensus_fasta': "TAREAN_consensus_rank_{}.fasta",
         'summary_histogram' : 'summary_histogram.png',
         'comparative_summary_map': 'comparative_summary.png',
         "how_to_cite" : "HOW_TO_CITE.html",
         'logfile' : "logfile.txt",
         'contigs' : "contigs.fasta",
         'filter_omitted' : DIRECTORY_TREE['sequences'] + "/removed_filtering_positive_reads.fasta",
         'filter_kept' : DIRECTORY_TREE['sequences'] + "/kept_filtering_positive_reads.fasta"
}


# include in output-  [source, destination]
INCLUDE = [
    [PATH("HOW_TO_CITE.html"), FILES["how_to_cite"]]
]

# this is attribute of path - not a file name!
FILES_TO_DISCARD_AT_CLEANUP = [
    'prerun', 'mgblast', 'blastn', "blastx",
    'hitsort', "repeat_annotation_summary_rds"
]

# relative links for html files
HTML_LINKS = {
    "CLUSTER_TO_SUPERCLUSTER" : "../../superclusters/dir_SC%04d/index.html",
    "SUPERCLUSTER_TO_CLUSTER" : "../../clusters/dir_CL%04d/index.html",
    "CLUSTER_TO_CLUSTER" : "../dir_CL%04d/index.html",
    "SUPERCLUSTER_TO_SUPERCLUSTER" : "../dir_SC%04d/index.html",
    "CLUSTER_TO_CLUSTER_TABLE" : "../../../../cluster_report.html",
    "SEPERCLUSTER_TO_CLUSTER_TABLE" : "../../../../cluster_report.html",
    "ROOT_TO_CLUSTER" : "seqclust/clustering/clusters/dir_CL%04d/index.html",
    "ROOT_TO_SUPERCLUSTER" : "seqclust/clustering/superclusters/dir_SC%04d/index.html",
    "ROOT_TO_TAREAN" : "seqclust/clustering/clusters/dir_CL%04d/tarean/report.html",
    "CLUSTER_TO_KMER_REPORT" : "tarean/report.html",
    "INDEX_TO_TAREAN": "tarean_report.html",
    "INDEX_TO_CLUSTER_REPORT": "cluster_report.html",
    "INDEX_TO_SUPERCLUSTER_REPORT" : "supercluster_report.html",
    "INDEX_TO_SUMMARIZED_ANNOTATION" : "summarized_annotation.html"
}


EMAX = 42.6  # define how many graph edges can be processed in 1Kb RAM
# MINIMUM_NUMBER_OF_INPUT_SEQUENCES = 5000
# FOR TESTING:
MINIMUM_NUMBER_OF_INPUT_SEQUENCES = 1000
MINIMUM_NUMBER_OF_READS_IN_CLUSTER = 20  # smaller clusters are not analyzed
MINIMUM_NUMBER_OF_READS_FOR_MERGING = 20  # smaller clusters will not be merged
MINIMUM_NUMBER_OF_SHARED_PAIRS_FOR_MERGING = 20  # min size of W param


NUMBER_OF_SEQUENCES_FOR_PRERUN_WITH_FILTERING = 50000
NUMBER_OF_SEQUENCES_FOR_PRERUN_WITHOUT_FILTERING = 20000
NUMBER_OF_SEQUENCES_FOR_PRERUN = NUMBER_OF_SEQUENCES_FOR_PRERUN_WITHOUT_FILTERING
CHUNK_SIZE = 20000


CLUSTER_EMAX = 2E7  # this parameter higle affect memory usage!

CLUSTER_VMAX = 40000
SUPERCLUSTER_THRESHOLD = 0.1
# Number of processors to use - it will be set at runtime
PROC = None
RSERVE_PORT = 6311

#some settings related to repeats annotation
ORF_THRESHOLD = 1200
PBS_THRESHOLD = 2
# threshold for rDNA detection - percentage of similarity hits
RDNA_THRESHOLD = 20
# threshold for contamination detection
CONTAMINATION_THRESHOLD = 10
# tandem ranks codes:
# 1 : putative tandem repeats - high confidence
# 2 : putative tandem repeats - low confidence
# 3 : potential LTR element
# 4 : rDNA
TANDEM_RANKS = [1, 2, 3, 4]
SKIP_CAP3_ASSEMBLY_TANDEM_RANKS = [1]
FILTER_MIN_PROP_THRESHOLD = 0.03   # this is minimal proportion of graph edges!
FILTER_MIN_SIZE_THRESHOLD = 1000  # minimal size of the cluster to be consider for filtering
FILTER_PROPORTION_OF_KEPT = 0.1

R = 'lib'
# external scripts
RSOURCE_tarean = PATH('lib/tarean/tarean.R')
RSOURCE_reporting = PATH('lib/reporting.R')
RSOURCE_create_annotation = PATH('lib/create_annotation.R')
LTR_DETECTION = PATH("lib/detect_LTR_insertion_sites.pl")

#PATH to DATABASES:
DNA_DATABASE = PATH("databases/dna_database_masked.fasta")
TRNA_DATABASE = PATH("databases/tRNA_database.fasta")
SATELLITE_MODEL = PATH("databases/satellite_model.rds")
LASTAL_PARAMS = PATH("databases/lastal_params")
# for testing
PROTEIN_DATABASE = None
CLASSIFICATION_HIERARCHY = None
CUSTOM_DNA_DATABASE = None

# when modifying this section check if makefile has most recent target for protein database
PROTEIN_DATABASE_DEFAULT = "VIRIDIPLANTAE3.0"
PROTEIN_DATABASE_OPTIONS = {'VIRIDIPLANTAE3.0' :
                            (PATH("databases/protein_database_viridiplantae_v3.0.fasta"),   # change according if you use custom protein database
                             PATH("databases/classification_tree_viridiplantae_v3.0.rds")),  # classification schem - data.tree object
                            'VIRIDIPLANTAE2.2' :
                            (PATH("databases/protein_database_viridiplantae_v2.2.fasta"),   # change according if you use custom protein database
                             PATH("databases/classification_viridiplantae_tree.rds")),  # classification schem - data.tree object
                            'METAZOA2.0' :
                            (PATH("databases/protein_database_metazoa_v3.fasta"), # change according if you use custom protein database
                             PATH("databases/classification_tree_metazoa_v3.rds")),   # classification schem - data.tree object
                            'METAZOA3.0' :
                            (PATH("databases/protein_database_metazoa_v3.fasta"), # change according if you use custom protein database
                             PATH("databases/classification_tree_metazoa_v3.rds"))   # classification schem - data.tree object
}
# if you change PROTEIN_DATABASE_OPTIONS, do not forget to use 'makeblastdb' build blast database
# and 'diamond makedb' to build diamond database

# PATH to binaries
LOUVAIN = PATH("louvain")
BINARIES = PATH("bin")

CAP3_PATTERNS_REPLACE = {"{}.{}.contigs" : [">Contig", ">{}Contig"],
                         "{}.{}.aln" : ["* Contig", "* {}Contig"],
                         "{}.{}.contigs.qual" : [">Contig", ">{}Contig"],
                         "{}.{}.ace" : ["CO Contig", "CO {}Contig"],
                         "{}.{}.singlets" : None,
                         "{}.{}.info" : None,
                         "{}.{}.contigs.links" : None}

CAP3_FILES_MAPPING = {"{}.{}.contigs" : "small_clusters.fasta",
                      "{}.{}.contigs.qual" : "small_clusters.contigs.qual",
                      "{}.{}.aln" : "small_clusters.aln",
                      "{}.{}.ace" : "small_clusters.ace",
                      "{}.{}.singlets" : None,
                      "{}.{}.info" : None,
                      "{}.{}.contigs.links" : None}

CAP3_FILENAMES = list(CAP3_PATTERNS_REPLACE.keys())

CAP3_FILES_GOODNAMES = {"{}.{}.contigs" : "contigs.fasta",
                        "{}.{}.contigs.qual" : "contigs.qual",
                        "{}.{}.aln" : "contigs.aln",
                        "{}.{}.ace" : "contigs.ace",
                        "{}.{}.singlets" : "singlets.fasta",
                        "{}.{}.info" : "assembly.info",
                        "{}.{}.contigs.links" : "contigs.links"}




CAP3_PARAMS = " -p 80 -o 40 " ## not implemented yet

LTR_DETECTION_FILES = {'ADJ': 'LTR_info.ADJ',
                       'LTR': 'LTR_info.LTR',
                       'PBS_BLAST': 'LTR_info.with_PBS_blast.csv',
                       'BASE' : 'LTR_info'}

# options for all-2-all search and annotations
FilteringThreshod = namedtuple("FilteringThreshold",
                               "min_lcov min_pid min_ovl min_scov evalue")
AnnotationParams = namedtuple("AnnotationParams", "blastn blastx blastn_trna")
Option = namedtuple('Options',
                    ('name database all2all_search_params filtering_threshold '
                     'filter_self_hits legacy_database lastdb  annotation_search_params'))

# protein domain search options:
DIAMOND = {
    'args': ' -p {max_proc} --max-target-seqs 1 --min-score 30 --freq-sd 1000 --more-sensitive',
    'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
    'column_types' : [str, str, float, float, float, float, float],
    'program': 'diamond blastx',
    'filter_function' : lambda x: x.bitscore >= 30,
    'parallelize' : False
}
BLASTX_W3 = {
    'args': ' -num_alignments 1 -word_size 2 -evalue 0.01 ',
    'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
    'column_types' : [str, str, float, float, float, float, float],
    'program': 'blastx',
    'filter_function' : lambda x: x.bitscore >= 33
}
BLASTX_W2 = BLASTX_W3
BLASTX_W2['args'] = ' -num_alignments 1 -word_size 3 -evalue 0.01 '


ARGS = None

ILLUMINA = Option(
    name="illumina",
    database='blastdb_legacy',
    all2all_search_params=('mgblast -p 75 -W18 -UT -X40 -KT -JF  -F '
                           '"m D" -v100000000 -b100000000'
                           ' -D4 -C 30 -H 30 -i {query} -d {blastdb}'),
    filtering_threshold=FilteringThreshod(55, 90, 0, 0, 1),
    filter_self_hits=False,
    legacy_database=True,
    lastdb=False,
    annotation_search_params=AnnotationParams(
        blastn={
            'args': ' -task blastn  -num_alignments 1 -evalue 0.01 ',
            'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
            'column_types' : [str, str, float, float, float, float, float],
            'program': 'blastn',
            'filter_function' : lambda x: x.length > 30 and x.bitscore > 60

        },
        blastx=BLASTX_W3,
        blastn_trna={
            'args': ' -task blastn  -num_alignments 1 -word_size 7',
            'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
            'column_types' : [str, str, float, float, float, float, float],
            'program': 'blastn',
            'filter_function' : lambda x: x.length > 18 and x.bitscore > 60
        }
    )
)

ILLUMINA_DUST_OFF = ILLUMINA._replace(
    all2all_search_params=('mgblast -p 75 -W18 -UT -X40 -KT -JF  -F '
                           'F -v100000000 -b100000000'
                           ' -D4 -C 30 -H 30 -i {query} -d {blastdb}'),
)


ILLUMINA_SENSITIVE_MGBLAST = ILLUMINA._replace(
    all2all_search_params=('mgblast -p 75 -W8 -UT -X40 -KT -JF  -F '
                           'F -v100000000 -b100000000'
                           ' -D4 -C 30 -H 30 -i {query} -d {blastdb}'),
    filtering_threshold=FilteringThreshod(55, 80, 0, 0, 1),
)



ILLUMINA_SENSITIVE_BLASTPLUS = ILLUMINA._replace(
    name="illumina_short",
    database="blastdb",
    all2all_search_params=('blastplus_wrapper.py blastn -task blastn  -word_size 6 -perc_identity 80 -max_hsps 100000000 -max_target_seqs 100000000 -xdrop_gap 40 -xdrop_ungap 40'
                           ' -outfmt \\"6 qseqid qlen qstart qend sseqid slen sstart send pident score evalue sstrand\\" -query {query} -db {blastdb}'),
    filtering_threshold=FilteringThreshod(55, 80, 0, 0, 0.1)
)


OXFORD_NANOPORE = Option(
    name="oxford_nanopore",
    database='lastdb',
    all2all_search_params=('last_wrapper.py -f blasttab+ -P1 '
                           ' -m 700 -p {} '
                           ' {{blastdb}} {{query}}  ').format(LASTAL_PARAMS),
    filtering_threshold=FilteringThreshod(55, 50, 0, 0, 0.01),
    filter_self_hits=True,
    legacy_database=False,
    lastdb=True,
    annotation_search_params=AnnotationParams(
        blastn={
            'args': ' -task blastn  -num_alignments 1 -evalue 0.01 -word_size 11',
            'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
            'column_types' : [str, str, float, float, float, float, float],
            'program': 'blastn',
            'filter_function' : lambda x: x.length > 30 and x.bitscore > 50
        },
        blastx={
            'args': ' -num_alignments 1 -word_size 2 -evalue 0.1',
            'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
            'column_types' : [str, str, float, float, float, float, float],
            'program': 'blastx',
            'filter_function' : lambda x: x.bitscore >= 30
        },
        blastn_trna={
            'args': ' -task blastn  -num_alignments 1 -word_size 7',
            'output_columns' : "qseqid sseqid qlen slen length ppos bitscore",
            'column_types' : [str, str, float, float, float, float, float],
            'program': 'blastn',
            'filter_function' : lambda x: x.length > 18 and x.length > 60
        }
    )
)
